Harden validation of user-submitted LCP background image URL

westonruter opened the pull request:
<!--
> [!IMPORTANT]
> This is a follow-up PR to #1697. Don't merge this until that PR is merged into `trunk`.
-->

Fixes #1584

Accepting the LCP background image URL from the untrusted client is dangerous as it could lead to abuse. The JSON Schema to enforce the validity of the URL Schema with such a background image URL is not robust enough to ensure that the URL is safe and that it actually references a valid image file. Therefore, a new `od_url_metric_data_pre_storage` filter is introduced in the REST API endpoint. The URL Metric data is passed into this filter, allowing plugins to further sanitize the data prior to storage. Alternatively, if the data cannot be sanitized due to it being invalid, an exception can be thrown (ideally an `OD_Data_Validation_Exception`) which blocks the URL Metric from being stored.

westonruter committed:
Revert "Remove od_store_url_metric_validity filter to be re-added in follow-up PR"

This reverts commit 514c51344f993c010edff305c27084c65e3da70d.

westonruter committed:
Use 'status' key instead of 'code'

Co-authored-by: felixarntz <flixos90@git.wordpress.org>

westonruter commented on plugins/image-prioritizer/helper.php:
```diff
@@ -136,6 +136,66 @@ function image_prioritizer_add_element_item_schema_properties( array $additional
+ * @param OD_Strict_URL_Metric $url_metric URL Metric, already validated against the JSON Schema.
+ * @return bool|WP_Error Validity. Valid if true or a WP_Error without any errors, or invalid otherwise.
+ */
+function image_prioritizer_filter_store_url_metric_validity( $validity, OD_Strict_URL_Metric $url_metric ) {
```
Copying thread from https://github.com/WordPress/performance/pull/1697#discussion_r1866290557:

@adamsilverstein:

> is part of the goal here to validate that the stored path actually returns an image? Is this something that could be checked on the client side?

@westonruter:

> The goal is to validate that the URL is valid. Since a malicious client could submit a URL to point to anything (e.g. even some 3P tracker URL), it's important to validate that (1) it's actually an image, (2) it's on a safe host.

westonruter commented on plugins/image-prioritizer/helper.php:
```diff
@@ -136,6 +136,66 @@ function image_prioritizer_add_element_item_schema_properties( array $additional
+	}
+
+	$r = wp_safe_remote_head(
+		$data['url'],
```
Copied from https://github.com/WordPress/performance/pull/1697#discussion_r1859375043:

The `wp_safe_remote_head()` function passes the supplied URL through `wp_http_validate_url()`, but that doesn't necessarily mean the request is safe. I'm thinking there should perhaps be an allowlist of hosts that this is allowed to connect to. Maybe we query for the first image attachment and pass it into `wp_get_attachment_image_url()` to then extract the hostname from the image. This would be better than having to try to come up with a list of all image CDNs that are used by plugins.

The list of allowed hosts could also be sent to the frontend so as to avoid sending the invalid URL in the first place. See related todo:

https://github.com/WordPress/performance/blob/bc81588943e835707d70948bdd6a3c817c948a25/plugins/optimization-detective/detect.js#L361-L366

Then again, would be a shape to continuously reject all of the URL Metrics in the case that an unexpected host is being used for a given background image. In that case, perhaps we should allow `OD_URL_Metric` to be mutable, so in the case where an invalid host is provided this callback could instead do:

```php
$url_metric->set( 'lcpElementExternalBackgroundImage', null );
```

or rather

```php
$url_metric->unset( 'lcpElementExternalBackgroundImage' );
```

When mutating an instance of `OD_URL_Metric` it should re-trigger the validation so that the mutation would cause the JSON Schema to be made invalid that it would throw an exception.

Also, whenever mutation happens then it should call a new `$this->group->clear_cache()` method which empties out `$group->result_cache` and which would in turn call `$collection->clear_cache()`.

So this `od_store_url_metric_validity` filter could be used in two ways:

1. A filter callback could return `false` or a non-empty `WP_Error` instance to block the URL Metric from being stored at all.
2. A filter callback could mutate the supplied `OD_URL_Metric` object to apply custom sanitizations or validation constraints which then would end up getting persisted in storage.

westonruter commented on plugins/optimization-detective/storage/rest-api.php:
```diff
@@ -210,6 +210,35 @@ function od_handle_rest_request( WP_REST_Request $request ) {
+	 * @param bool|WP_Error        $validity   Validity. Valid if true or a WP_Error without any errors, or invalid otherwise.
+	 * @param OD_Strict_URL_Metric $url_metric URL Metric, already validated against the JSON Schema.
+	 */
+	$validity = apply_filters( 'od_store_url_metric_validity', true, $url_metric );
```
Copying from https://github.com/WordPress/performance/pull/1697#discussion_r1865244837:

To compliment https://github.com/WordPress/performance/pull/1697/files#r1859381128, this would allow a filter callback to always have access to whatever the original data was, regardless of any changes which have been made by other callbacks to the `$url_metric` object. This would be analogous to the second `$postarr` argument to the `wp_insert_post_data` filter, after the filtered `$data` arg.

```suggestion
	 * @param bool|WP_Error        $validity        Validity. Valid if true or a WP_Error without any errors, or invalid otherwise.
	 * @param OD_Strict_URL_Metric $url_metric      URL Metric, already validated against the JSON Schema.
	 * @param array<string, mixed> $url_metric_data Original URL Metric data before any mutations.
	 */
	$validity = apply_filters( 'od_store_url_metric_validity', true, $url_metric, $url_metric->jsonSerialize() );
```

westonruter committed:
Merge branch 'trunk' of https://github.com/WordPress/performance into add/external-bg-preload-validation

westonruter committed:
Add clear_cache() method to OD_URL_Metric_Group

westonruter committed:
Add ability to unset an extended property on OD_URL_Metric and OD_Element

adamsilverstein commented on plugins/image-prioritizer/helper.php:
```diff
@@ -137,6 +137,66 @@ function image_prioritizer_add_element_item_schema_properties( array $additional
+		return $validity;
+	}
+
+	$r = wp_safe_remote_head(
```
Check that `$data['url']` exists before using (next line)?

westonruter replied:
It should be guaranteed to exist because the extended JSON Schema is enforcing that the `lcpElementExternalBackgroundImage` key will have the specific shape.

westonruter replied:
I went ahead and implemented this in 8f2af8766f6ef99e66cc7db634c674352a913f6d

adamsilverstein commented on plugins/image-prioritizer/helper.php:
```diff
@@ -137,6 +137,66 @@ function image_prioritizer_add_element_item_schema_properties( array $additional
+		);
+	}
+
+	// TODO: Check for the Content-Length and return invalid if it is gigantic?
```
plus maybe a filter for setting the threshold :)

westonruter replied:
Maybe, but maybe we can make a decision about what the maximum size of a background image should be that we would possibly want to preload. If the 90th percentile largest image on a desktop page is 1MB [according to Web Almanac 2022](https://almanac.httparchive.org/en/2022/media#fig-12), maybe we should make 2MB the maximum that we would preload?

westonruter replied:
Implemented in bcefd693e3dc4944681418a6f8865220f397f0b7

adamsilverstein commented on plugins/optimization-detective/tests/test-class-od-element.php:
```diff
@@ -130,22 +131,68 @@ static function ( array $schema ): array {
+				$element->offsetSet( 'isLCP', true );
+			},
+		);
+		foreach ( $functions as $function ) {
```
did you consider having the arrays of functions in this test come from data providers?

westonruter replied:
That could work, but here I just wanted to test different flavors of a specific aspect of something, so adding a data provider to test everything with this one variation would I think be overkill.

adamsilverstein commented on plugins/optimization-detective/tests/test-class-od-url-metrics-group-collection.php:
```diff
@@ -199,6 +199,89 @@ public function data_provider_sample_size_and_breakpoints(): array {
+				return $schema;
+			}
+		);
+		add_filter(
```
all filters are removed after each test automatically, right? i often try to clean up any filters I add in tests out of habit which I believe is superfluous.

westonruter replied:
Yes, filters are automatically reset after each test.

westonruter replied:
https://github.com/WordPress/wordpress-develop/blob/e2b0bfdc06b2f652ff7b3865b496325b87521d26/tests/phpunit/includes/abstract-testcase.php#L202C10-L202C24

westonruter committed:
Suppress erroneous IDE warnings

westonruter committed:
Unset lcpElementExternalBackgroundImage if URL is invalid

westonruter committed:
Merge branch 'trunk' of https://github.com/WordPress/performance into add/external-bg-preload-validation

westonruter committed:
Improve docs and tidiness

westonruter committed:
Add tests for od_url_metric_storage_validity filter

westonruter committed:
Fix typo in readme

westonruter commented on plugins/optimization-detective/storage/rest-api.php:
```diff
@@ -210,14 +210,81 @@ function od_handle_rest_request( WP_REST_Request $request ) {
+		if ( false === $validity ) {
+			$validity = new WP_Error( 'invalid_url_metric', __( 'Validity of URL Metric was rejected by filter.', 'optimization-detective' ) );
+		}
+		$error_code = $validity->get_error_code();
+		if ( ! isset( $validity->error_data[ $error_code ]['status'] ) ) {
+			$validity->error_data[ $error_code ]['status'] = 400;
+		}
```
I need to double check this is the right way to do this.

westonruter replied:
All of the tests indicate this is right.

westonruter commented:
@felixarntz This is ready for review although I still have yet to:

- [x] Guard against attempting to preload gigantic images (e.g. 2MB).
- [x] Add tests for functions added in `plugins/image-prioritizer/helper.php`.

I intend to do these Friday morning my time, so hopefully this can be reviewed and merged by mid-day Friday PST in preparation for the Monday release.

westonruter commented:
I ran out of time to work on this today, unfortunately. So I'm going to pick it up tomorrow.

westonruter committed:
Scaffold new tests for Image Prioritizer

westonruter committed:
Merge branch 'trunk' of https://github.com/WordPress/performance into add/external-bg-preload-validation

westonruter committed:
Add missing access private tags

westonruter committed:
Add tests for various validity conditions for external BG images

westonruter committed:
Fix handling of invalid external BG image and add tests

westonruter committed:
Avoid preloading background images larger than 2MB

westonruter committed:
Update readme to relate features of Image Prioritizer

felixarntz commented on plugins/optimization-detective/class-od-element.php:
```diff
@@ -208,18 +208,53 @@ public function offsetSet( $offset, $value ): void {
+	 *
+	 * @throws OD_Data_Validation_Exception When attempting to unset a property required by the JSON Schema.
+	 */
+	public function unset( string $key ): void {
```
I'm not a fan of introducing an `unset()` method as part of the OD API, as that's not something we typically want to encourage.

Since almost all of its usage is for tests (which we could work around with `ReflectionProperty`), I wonder whether we can come up with another solution for this?

felixarntz commented on plugins/optimization-detective/class-od-url-metric.php:
```diff
@@ -503,6 +503,33 @@ function ( array $element ): OD_Element {
+	 * @param string $key Key to unset.
+	 * @throws OD_Data_Validation_Exception If the property is required an exception will be thrown.
+	 */
+	public function unset( string $key ): void {
```
See above.

felixarntz commented on plugins/image-prioritizer/helper.php:
```diff
@@ -137,14 +142,173 @@ function image_prioritizer_add_element_item_schema_properties( array $additional
+			 * @noinspection PhpUnhandledExceptionInspection
+			 */
+			wp_trigger_error( __FUNCTION__, $image_validity->get_error_message() . ' Background image URL: ' . $data['url'] );
+			$url_metric->unset( 'lcpElementExternalBackgroundImage' );
```
It looks like this is the only `unset()` usage in production code? Since this is a very specific use-case, maybe we can find another solution to make it possible, without allowing to unset metric and element data generally.

Could we for instance generate a copy of the given `$url_metric`, without the field that is unset? Obviously that's more or less the same thing, but without the need for an extra method just because of this use-case.

westonruter replied:
That would work. Such a sanitization filter would be something like this:

```php
$data = $url_metric->jsonSerialize();
unset( $data['lcpElementExternalBackgroundImage'] );
$url_metric = new OD_Strict_URL_Metric( $data );
```

The downside here is it's going to end up re-validating the data against the schema twice.

felixarntz replied:
That's probably fine. I don't think there's a considerable performance cost with that, and it only happens when sending data (vs e.g. on every page load).

felixarntz commented on plugins/optimization-detective/storage/rest-api.php:
```diff
@@ -210,14 +210,81 @@ function od_handle_rest_request( WP_REST_Request $request ) {
+		 * @param OD_Strict_URL_Metric $url_metric      URL Metric, already validated against the JSON Schema.
+		 * @param array<string, mixed> $url_metric_data Original URL Metric data before any mutations.
+		 */
+		$validity = apply_filters( 'od_url_metric_storage_validity', true, $url_metric, $url_metric->jsonSerialize() );
```
Is this really storage specific? Why not call it just `od_url_metric_validity`?

westonruter replied:
The rationale for this being storage-specific explained in the filter description. In particular, this extra validation is only applied at the time of storage. It doesn't apply when loading URL Metrics from the `od_url_metrics` post type.

felixarntz replied:
I think that's quite normal though, to validate and sanitize prior to receiving data. When loading from a post type, it's fair to assume that this data is valid - otherwise it should have never been saved to a post in the first place.

I don't think WordPress commonly validates or sanitizes outside of receiving data, so I think mentioning "storage" seems unnecessary.

felixarntz commented on plugins/image-prioritizer/helper.php:
```diff
@@ -137,14 +142,173 @@ function image_prioritizer_add_element_item_schema_properties( array $additional
+	if ( is_array( $data ) && isset( $data['url'] ) && is_string( $data['url'] ) ) { // Note: The isset() and is_string() checks aren't necessary since the JSON Schema enforces them to be set.
+		$image_validity = image_prioritizer_validate_background_image_url( $data['url'] );
+		if ( is_wp_error( $image_validity ) ) {
+			/**
+			 * No WP_Exception is thrown by wp_trigger_error() since E_USER_ERROR is not passed as the error level.
+			 *
+			 * @noinspection PhpUnhandledExceptionInspection
+			 */
+			wp_trigger_error( __FUNCTION__, $image_validity->get_error_message() . ' Background image URL: ' . $data['url'] );
+			$url_metric->unset( 'lcpElementExternalBackgroundImage' );
```
Unrelated to the `unset()` point, this function strictly speaking sanitizes the URL metric, it doesn't just validate it. Given the validity filter name, I would find it more appropriate to throw an exception and call it a day. Validation shouldn't remove invalid properties. Also, since this `wp_trigger_error()` would typically happen in the REST API, it would be suppressed even with `WP_DEBUG_DISPLAY`, making this harder to spot - so an exception / `WP_Error` would be better from that perspective too. Would that not work?

westonruter replied:
My thinking behind unset here is that what if someone using an external background image from another domain not in the allowed list of origins. If any exception it thrown then every single URL Metric will be rejected. This is not good because then none of the other data in the URL Metric would ever be stored. What we want in this case is to just exclude that non-required property from the URL Metric as if it wasn't set in the first place. The use of `wp_trigger_error()` is to provide a mechanism for a site owner to find out when `WP_DEBUG` is enabled.

felixarntz replied:
That makes sense, but then shouldn't we have a filter for sanitization instead for this purpose, rather than validation?

Given this is for now the only usage of the validation filter, but it's not throwing an exception, I think it would make sense to change this to a sanitization filter that acts on the `URL_Metric` instance. This would be similar to e.g. how Core has a sanitization filter for options, but not validation. Obviously we could also add a validation extension point, but that could then happen later and doesn't seem as relevant to this as a sanitization filter.

westonruter replied:
Here's an alternative to the current validity filter in this PR. We could have a filter that runs on the `$data` before it is passed into the `OD_Strict_URL_Metric` class for instantiation. So instead of:

https://github.com/WordPress/performance/blob/6005f4a9f084825edc4f2eeeb4152b41b29ef0ce/plugins/optimization-detective/storage/rest-api.php#L188-L200

We could have:

```php
try {
	$data = array_merge(
		$data,
		array(
			// Now supply the readonly args which were omitted from the REST API params due to being `readonly`.
			'timestamp' => microtime( true ),
			'uuid'      => wp_generate_uuid4(),
			'etag'      => $request->get_param( 'current_etag' ),
		)
	);

	/**
	 * Filters the URL Metric data prior to validation for storage.
	 *
	 * This can be used to apply additional sanitization to the URL Metric data which
	 * cannot be done with JSON Schema alone. Additionally, this filter can be used to
	 * apply validation constraints by throwing an OD_Data_Validation_Exception which
	 * will block the URL Metric from being stored.
	 *
	 * @since n.e.x.t
	 * @param array<string, mixed> $data URL Metric data. This is the Data type from OD_URL_Metric.
	 */
	$data = apply_filters( 'od_url_metric_data_pre_storage', $data );

	// The "strict" URL Metric class is being used here to ensure additionalProperties of all objects are disallowed.
	$url_metric = new OD_Strict_URL_Metric( $data );
```

In this way, the same filter can be used for both sanitization and validation. Even without explicitly throwing an exception, an exception would end up getting thrown anyway if a filter did something like `unset( $data['viewport] )` since that is a required property.

westonruter replied:
See ~94c5e6b~ ~e6cfcfdb0~ 250f094c1

felixarntz replied:
@westonruter That looks good to me. Though I think that filter name would be better as `od_sanitize_url_metric_data`, as it's a more intuitive name in terms of naming structure, and its primary purpose would be to sanitize data.

Makes sense that this filter could also be used for validation, so the `try catch` wrap around it makes sense.

westonruter replied:
I don't think using "sanitize" in the filter name makes sense since other core filters aren't named in that way. It's up to the filter callback to decide whether the purpose is to sanitize. Going the Options API as a model, there is the `pre_update_option` filter which is used to (maybe) serialize, how about `od_pre_store_url_metric`? Or to follow `wp_insert_post()` which has the `wp_insert_post_data` filter, what about `od_store_url_metric_data`? I like that more.

westonruter replied:
Well, there there is `sanitize_option_{$option_name}`.

felixarntz replied:
> I don't think using "sanitize" in the filter name makes sense since other core filters aren't named in that way.

I don't think that's true, see e.g. `sanitize_option_{$option}`.

> It's up to the filter callback to decide whether the purpose is to sanitize. Going the Options API as a model, there is the `pre_update_option` filter which is used to (maybe) serialize, how about `od_pre_store_url_metric`?

What other use-cases would such a filter have than sanitize? You can throw an exception (effectively validate), but technically you can throw an exception from anywhere, so that to me is not necessarily a reason to change the name. The example of serialization wouldn't be relevant here.

> Or to follow `wp_insert_post()` which has the `wp_insert_post_data` filter, what about `od_store_url_metric_data`? I like that more.

This would be an okay name to me if we go with filtering the raw data. But that's not as clean of an API surface as filtering the metric instance (per my other comment), so I would prefer not to do that.

westonruter replied:
> What other use-cases would such a filter have than sanitize? You can throw an exception (effectively validate), but technically you can throw an exception from anywhere, so that to me is not necessarily a reason to change the name. The example of serialization wouldn't be relevant here.

A use case could be an extension which amends a stored URL Metric with additional data, like the IP address of the client which submitted the URL Metric.

> This would be an okay name to me if we go with filtering the raw data. But that's not as clean of an API surface as filtering the metric instance (per my other comment), so I would prefer not to do that.

I think the API surface is actually cleaner by passing around an array rather then the `OD_URL_Metric` instance. See https://github.com/WordPress/performance/pull/1713#discussion_r1887354549.

felixarntz reviewed (changes requested):
@westonruter The actual validation logic in Image Prioritizer looks good, but a few points of feedback on the OD API implementation.

westonruter committed:
Replace validity filter with sanitization filter; unset unset()

felixarntz commented on plugins/optimization-detective/storage/rest-api.php:
```diff
@@ -186,38 +186,84 @@ function od_handle_rest_request( WP_REST_Request $request ) {
+		 *
+		 * @param array<string, mixed> $data URL Metric data. This is the Data type from OD_URL_Metric.
+		 */
+		$data = apply_filters( 'od_url_metric_data_pre_storage', $data );
```
See https://github.com/WordPress/performance/pull/1713#discussion_r1887336707 and https://github.com/WordPress/performance/pull/1713#discussion_r1887338680, I think there are better names for such a filter. Two things:
* It's missing a verb. I think `sanitize` would be suitable.
* The mentioning of "pre storage" is unnecessary for this kind of sanitization (or validation), and I think it's unnecessarily verbose.

westonruter replied:
I've renamed the filter in f74f4f4b1b08dfb71b4eb0fc25d03e8859dad330

felixarntz commented on plugins/optimization-detective/storage/rest-api.php:
```diff
@@ -186,38 +186,84 @@ function od_handle_rest_request( WP_REST_Request $request ) {
+		 *
+		 * @param array<string, mixed> $data URL Metric data. This is the Data type from OD_URL_Metric.
+		 */
+		$data = apply_filters( 'od_url_metric_data_pre_storage', $data );
+
+		// The "strict" URL Metric class is being used here to ensure additionalProperties of all objects are disallowed.
+		$url_metric = new OD_Strict_URL_Metric( $data );
```
Can we not keep filtering the `OD_URL_Metric` instance for this, as that would be a better filter interface than dealing with raw data?

westonruter replied:
I'd rather not because then we're passing around an object to the filters, and objects are passed by reference. And any time that you construct an `OD_URL_Metric` it applies all of the JSON Schema for validation. It's also not ergonomic to make a change to a URL Metric via a filter:

```php
function ( $url_metric ) {
    $data = $url_metric->jsonSerialize();
    unset( $data['bad_key'] );
    $new_url_metric = new OD_Strict_URL_Metric( $data );
    return $new_url_metric;
}
```

Compared with:

```php
function ( $data ) {
    unset( $data['bad_key'] );
    return $data;
}
```

So I think it is better to pass around the URL Metric data

westonruter replied:
Also, by passing around an object then the user needs to be aware of the distinction between `OD_Strict_URL_Metric` and `OD_URL_Metric`, where the former is intended to be an internal class used just in the REST API endpoint.

felixarntz replied:
That's only for the specific example of unsetting data though (because there's no `unset()` method). Regular sanitization would be as simple as:
```php
function ( $url_metric ) {
    $url_metric->set( 'prop', 'sanitized-value' );
    return $new_url_metric;
}
```

If you feel strongly about the ergonomics of this for the scenario of unsetting a property, I'd rather bring back the `unset()` method than filtering raw data just because of the lack of an `unset()` method.

Honestly, at that point, it would be even better to use an action hook, since the object can be modified directly either way, and an exception can still be thrown too, so a filter is not needed. This would also be safer, as a filter would allow to exchange the entire object which, despite me having suggested that earlier in favor of `unset()`, also has drawbacks as normally people shouldn't be doing that.

westonruter replied:
There is no `set()` method either. So you'd have to do the same:

```php
function ( $url_metric ) {
    $data = $url_metric->jsonSerialize();
    $data['prop'] = 'sanitized-value';
    $new_url_metric = new OD_Strict_URL_Metric( $data );
    return $new_url_metric;
}
```

If we were to support `set()` then this would bring back concerns about cache invalidation for the group and the collection, and it would be confusing if a user tries to call `set()` on a URL Metric from a collection in a tag visitor, as obviously that would not be right since it's only relevant here.

So I think it's preferable for the `OD_URL_Metric` object to be immutable, siding with you for suggesting that we don't allow `unset()`.

felixarntz replied:
Oh I must have confused seeing `OD_Element::offsetSet()`.

Taking a step back: The issues of passing around raw array data are that there is little control over it, no way to insert any kind of "middleware", plus no good property autocomplete support etc. So I think it's worth finding a better solution, and the current constraints of `OD_URL_Metric` and `OD_Strict_URL_Metric` don't seem right, based on the points you made.

How about something like this?
* We introduce another kind of URL metric class. This class would represent a URL metric without any validation in place and would only be used while a URL metric is being recorded.
* This class would have a set method, and could also have an unset method, as modifying what's recorded is crucial as part of what we're trying to achieve here. It would be passed to the new action (or filter) and allow ergonomically to customize what's being recorded, without any constraints of schema validation.
* After the hook, an instance of the current URL metric class is created from the "unvalidated" instance, so at that point validation fires and would catch any wrongdoing that might have happened during the hook.

A few architectural alternatives on how this could be implemented:
1. We introduce something like `OD_Unvalidated_URL_Metric` and have `OD_URL_Metric` extend it. Downside: A bit odd naming pattern with the negation of "validated" in the base class.
2. We name the new unvalidated class `OD_URL_Metric` and rename the current `OD_URL_Metric` to something like `OD_Validated_URL_Metric`, and have it extend the new `OD_URL_Metric`. Downside: A bit of a strange pattern would be required to override the set and unset methods to throw an exception instead of actually setting - not great to have methods on a class that shouldn't be there.
3. We introduce the new class separately from the existing URL metric classes, i.e. no new class extension. This way only the new class would have set and unset methods. Maybe this could be called something like `OD_Raw_URL_Metric`? Then either that class would have a method to return a validated URL metric instance from its data, or the current `OD_URL_Metric` would have a static method like `create_from_raw()` (or similar).
    * For this approach, ideally there should be an interface that both the new `OD_Raw_URL_Metric` and the current `OD_URL_Metric` would implement, to make sure the common part of their implementations remain aligned. If we go with an interface, maybe the interface itself should be named `OD_URL_Metric` and the current `OD_URL_Metric` class should then be renamed to something more specific, like `OD_Validated_URL_Metric`.

If we go with the interface or a renaming of the current `OD_URL_Metric` class, I realize this would be a bit of refactoring, so not ideal to do so shortly before the release as part of this PR. But once we agree on an approach, we could in this PR simply do the bare minimum to work towards it and address the rest in a follow up. For instance, if we decide to rename `OD_URL_Metric`, we don't have to do it as part of this PR. The main thing IMO is that we can discuss and decide on an option here, without thinking about the constraints of this PR too much, and then do the least effort needed towards that goal in this PR.

westonruter replied:
> The issues of passing around raw array data are that there is little control over it, no way to insert any kind of "middleware", plus no good property autocomplete support etc.

The filters are the middleware. Property auto completion is not currently supported anyway for custom properties. Only core properties could be possible autocompleted, and this can be implemented by a plugin importing the PHPStan `Data` type from  the `OD_URL_Metric` class.

> How about something like this?

I really think this is too much. I think it is somewhat over-engineered. What is in place right now feels WordPressy with a simple filter to process the raw data (which would have been pre-validated with the JSON schema attached to the REST API args), and then will be validated again anyway when the `OD_Strict_URL_Metric` is constructed. The filter in place now is similar to the `wp_insert_post_data` filter in WordPress so it's easy to understand and use.

If we want something more structured in the future I suggest deprecating this filter to instead go with the replacement. But I really think this current approach is best, especially for now.

felixarntz replied:
> The filters are the middleware.

Not in the sense that I meant it, they don't allow controlling what you can modify. With an array you can do whatever you want. If we later renamed a property and handle it backwards compatibly, or wanted to trigger a warning if someone changes a specific property, things like that, we can't do that if we use an array - but we can if we use a class instance.

> I really think this is too much. I think it is somewhat over-engineered.

Can you clarify why? What's "WordPressy" is a bit subjective, and FWIW some parts of WordPress are not necessarily examples for good engineering.

> If we want something more structured in the future I suggest deprecating this filter to instead go with the replacement.

That's a short-term option, though it would not be a good long-term solution IMO per my previous comment. Since this is in the milestones due tomorrow, I'd be okay going with it for now, but it's really not great to ship new API surfaces without sufficient consideration just because we want to ship something fast.

westonruter replied:
> If we later renamed a property and handle it backwards compatibly, or wanted to trigger a warning if someone changes a specific property, things like that, we can't do that if we use an array - but we can if we use a class instance.

That's a good point. 

> Can you clarify why?

It's subjective 😀

> What's "WordPressy" is a bit subjective

I'm not sure its subjective. WordPressy is what aligns with previous established patterns in WordPress.

> FWIW some parts of WordPress are not necessarily examples for good engineering.

This is true.

----

I think there is another way to go about this with an existing filter in the REST API without introducing a new sanitization/validation mechanism: the `rest_request_brefore_callbacks` filter. I've never used this filter before, but it is billed as being specifically for performing additional validation. I'm investigating.

westonruter replied:
Here, check this out: d4c9f40fb3307d11b1679da76c2e8333a80f91d0

felixarntz replied:
Sure, if we can implement this without a new API surface in OD entirely, that would simplify things, at least for now :)

felixarntz replied:
Thanks, that looks great and elegantly avoids expanding the new API. So obviously this is a workaround now, but it gives us time to think whether and how we would want to make this more intuitively possible.

felixarntz reviewed (commented):
@westonruter This looks close, two follow up comments.

westonruter committed:
Rename filter

swissspidy commented on plugins/image-prioritizer/helper.php:
```diff
@@ -137,14 +142,170 @@ function image_prioritizer_add_element_item_schema_properties( array $additional
+			 *
+			 * @noinspection PhpUnhandledExceptionInspection
+			 */
+			wp_trigger_error( __FUNCTION__, $image_validity->get_error_message() . ' Background image URL: ' . $data['lcpElementExternalBackgroundImage']['url'] );
```
Other error messages are translated, why not this one?

westonruter replied:
I guess because I was being lazy.

westonruter replied:
How about:

```suggestion
			wp_trigger_error(
				__FUNCTION__,
				$image_validity->get_error_message() . ' ' . __( 'Background image URL:', 'image-prioritizer' ) . $data['lcpElementExternalBackgroundImage']['url']
			);
```

swissspidy replied:
Ideal would be sprintf with a placeholder and a translator comment, no string concatenation.

So maybe something like

```php
sprintf(
	/* translators: 1: error message. 2: image url */
	__( 'Error: %1$s. Background image URL: %2$s., 'image-prioritizer' ),
	$image_validity->get_error_message(),
	$data['lcpElementExternalBackgroundImage']['url']
)
```

westonruter replied:
See e1d0ac9dd935634b782d711c7e1ae85d296f44cf

swissspidy commented on plugins/image-prioritizer/helper.php:
```diff
@@ -137,14 +142,170 @@ function image_prioritizer_add_element_item_schema_properties( array $additional
+	);
+
+	// Obtain the host of an image attachment's URL in case a CDN is pointing all images to an origin other than the home or site URLs.
+	$image_attachment_query = new WP_Query(
+		array(
+			'post_type'      => 'attachment',
+			'post_mime_type' => 'image',
+			'post_status'    => 'inherit',
+			'posts_per_page' => 1,
+			'fields'         => 'ids',
+		)
+	);
+	if ( isset( $image_attachment_query->posts[0] ) && is_int( $image_attachment_query->posts[0] ) ) {
+		$src = wp_get_attachment_image_src( $image_attachment_query->posts[0] );
+		if ( is_array( $src ) ) {
+			$attachment_image_src_host = wp_parse_url( $src[0], PHP_URL_HOST );
+			if ( is_string( $attachment_image_src_host ) ) {
+				$allowed_hosts[] = $attachment_image_src_host;
+			}
+		}
+	}
```
Worth caching this result?

Or could the WP_Query be optimized further, e.g. with `update_post_term_cache => false`?

westonruter replied:
How about 6856026e287becca40d4bca887fb73aef5427618?

westonruter committed:
Further optimize WP_Query

Co-authored-by: swissspidy <swissspidy@git.wordpress.org>

westonruter committed:
Improve translatability of error message

Co-authored-by: swissspidy <swissspidy@git.wordpress.org>

westonruter committed:
Update readme with links to where the performance features are implemented

westonruter committed:
Merge branch 'trunk' into add/external-bg-preload-validation

westonruter committed:
Eliminate od_store_url_metric_data filter in favor of reusing rest_request_before_callbacks

westonruter commented on plugins/optimization-detective/storage/rest-api.php:
```diff
@@ -174,7 +174,7 @@ function od_handle_rest_request( WP_REST_Request $request ) {
 	}
 
-	$data = $request->get_json_params();
+	$data = $request->get_json_params(); // TODO: Why not just get_params()?
```
I remember. It's because any URL query parameter would then be included among the params, so by just getting the params from the JSON body we avoid potential failures due to additional query parameters being injected into the request URL.

```suggestion
	$data = $request->get_json_params();
```

westonruter committed:
Remove todo

This is using get_json_params() instead of get_params() because any URL query parameter would then be included among the params, so by just getting the params from the JSON body we avoid potential failures due to additional query parameters being injected into the request URL.

westonruter committed:
Account for route matching being case insensitive

felixarntz reviewed (approved):
Thanks for the iterations @westonruter!

westonruter committed:
Improve function description and further trim route

westonruter commented on plugins/image-prioritizer/helper.php:
```diff
@@ -137,14 +142,193 @@ function image_prioritizer_add_element_item_schema_properties( array $additional
+		$request->get_method() !== 'POST'
+		||
+		// The strtolower() is due to \WP_REST_Server::match_request_to_handler() using case-insensitive pattern match.
+		OD_REST_API_NAMESPACE . OD_URL_METRICS_ROUTE !== trim( strtolower( trim( $request->get_route(), '/' ) ) )
```
By chance I had the thought of checking whether the route was case sensitive. It wasn't. So this would have been a vulnerability. I want to double check how `$request->get_route()` gets populated to see if there could potentially be any other ways to make a request to the endpoint and yet bypass this condition here.

westonruter replied:
Indeed, I found that if you tried making a request without pretty permalinks and added `%0A` to the end of the `rest_route` query parameter:

```
http://localhost:8888/index.php?rest_route=/optimization-detective/v1/url-metrics:store%0A
```

Then the route is parsed as `"/optimization-detective/v1/url-metrics:store\n"` and this matches here in `\WP_REST_Server::match_request_to_handler()` on [this line](https://github.com/WordPress/wordpress-develop/blob/c697356c29effcfa8115d852bc89385f7112c202/src/wp-includes/rest-api/class-wp-rest-server.php#L1171):

```php
$match = preg_match( '@^' . $route . '$@i', $path, $matches );
```

This is because `$` matches a newline. In reality this should be using `\z` instead.

westonruter replied:
I added a test case to explicitly check for this: https://github.com/WordPress/performance/pull/1713/commits/5ab7fd1a1ca07413795332a89fcf6bb3995360bb

westonruter committed:
Add test case for route ending in newline
